### PR TITLE
Add support for jackson field ids

### PR DESCRIPTION
**Pull request by brenbar**
### Background
Jackson core has interfaces for [field ids](https://github.com/FasterXML/jackson-core/blob/e4ac4f85873dc5e9cad8201f4b5b659db03b1f73/src/main/java/com/fasterxml/jackson/core/JsonGenerator.java#L1140-L1157). This is a great opportunity for msgpack, since the protocol allows for integer keys, enabling more advanced binary serialization strategies for further reduced message size.

Current implementation of msgpack-jackson only allows for coercing strings to integers. Implementing the formal interfaces will enable end-to-end map serialization with mixed string/integer keys. [Other msgpack implementations](https://github.com/msgpack/msgpack-python?tab=readme-ov-file#security) already support this.

### Summary of Changes
1. Implement the formal jackson core interface for writing integer keys as field ids instead of strings.
2. Add a helper function on the msgpack parser that can be used in a jackson `KeyDeserializer` to deserialize field ids.

### CODE REVIEW

File: msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java [hunk 1]
```diff
@@ -513,6 +513,13 @@ private void writeByteArrayTextKey(byte[] text, int offset, int len) throws IOEx
         addValueNode(new String(text, offset, len, DEFAULT_CHARSET));
     }
 
+    // TODO: Uncomment
+    //@Override
+    //public void writeFieldId(long id) throws IOException
+    //{
+    //    addKeyToStackTop(id);
```
Comment (brenbar) [inline msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java:520]: ☝️ Correction. Comment was drafted for older msgpack.
```suggestion
    //    addKeyNode(id);
```

**Comment by brenbar** [comment]
@komamitsu @xerial Can CI be run on this PR to show the failing test? Then I'll follow-up with the implementation changes.

**Comment by brenbar** [comment]
@komamitsu @xerial Another CI run please 🙏

**Comment by brenbar** [comment]
🔨 Build failed with test showcasing new edge case.
```
[info] Test org.msgpack.jackson.dataformat.MessagePackDataformatForFieldIdTest.testMixedKeys started
[error] Test org.msgpack.jackson.dataformat.MessagePackDataformatForFieldIdTest.testMixedKeys failed: java.lang.AssertionError: expected: java.util.HashMap<{1=one, 2=two}> but was: java.util.HashMap<{1=one, 2=two}>, took 0.258 sec
[error]     at org.msgpack.jackson.dataformat.MessagePackDataformatForFieldIdTest.testMixedKeys(MessagePackDataformatForFieldIdTest.java:105)
[error]     ...
[info] Test run org.msgpack.jackson.dataformat.MessagePackDataformatForFieldIdTest finished: 1 failed, 0 ignored, 1 total, 0.264s
```

Will now demonstrate passing test with implementation changes.

File: msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java [hunk 2]
```diff
@@ -513,6 +513,12 @@ private void writeByteArrayTextKey(byte[] text, int offset, int len) throws IOEx
         addValueNode(new String(text, offset, len, DEFAULT_CHARSET));
     }
 
+    @Override
+    public void writeFieldId(long id) throws IOException
+    {
+        addKeyNode(id);
```
Comment (brenbar) [inline msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java:519]: @komamitsu @xerial For backwards compatibility, I imagine we need some sort of feature flag that defaults to the prior implementation. Do you guys have any suggestions for how that flag should be defined? Should I just use the [`MessagePack.PackerConfig`](https://github.com/msgpack/msgpack-java/blob/59219838ffb6236eca84601dad1dd267541c070b/msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java#L52), or should I pass it [directly from the `MessagePackFactory` to the `MessagePackGenerator`](https://github.com/msgpack/msgpack-java/blob/59219838ffb6236eca84601dad1dd267541c070b/msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackFactory.java#L87)?
```suggestion
        if (writeIntegerMapKeysAsStringKeys) {
            super.writeFieldId(id);
        } else {
            addKeyNode(id);
        }
```

Comment (komamitsu) [reply msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java:519]: Sorry, for the delayed reply. `PackerConfig` is for the `core` configuration and shouldn't be used for `jackson-dataformat-msgpack`. I think it's okay to pass the flag like `reuseResource`.

File: msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java [hunk 3]
```diff
@@ -210,6 +212,24 @@ public MessagePackGenerator(
         this.messagePacker = packerConfig.newPacker(getMessageBufferOutputForOutputStream(out, reuseResourceInGenerator));
         this.packerConfig = packerConfig;
         this.nodes = new ArrayList<>();
+        this.writeIntegerKeysAsStringKeys = true;
```
Comment (komamitsu) [inline msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java:215]: I don't think this constructor is used now. Can you remove this?

File: msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java [hunk 4]
```diff
@@ -195,6 +196,7 @@ private MessagePackGenerator(
         this.messagePacker = packerConfig.newPacker(out);
         this.packerConfig = packerConfig;
         this.nodes = new ArrayList<>();
+        this.writeIntegerKeysAsStringKeys = true;
```
Comment (komamitsu) [inline msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java:199]: This constructor is called at https://github.com/msgpack/msgpack-java/blob/8247f243c4c9141b237c815caf4c4838e7a6bb64/msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java#L396.  `this.writeIntegerKeysAsStringKeys` needs to be passed to the constructor to use the same configuration. I think a new argument needs to be added.

Comment (brenbar) [reply msgpack-jackson/src/main/java/org/msgpack/jackson/dataformat/MessagePackGenerator.java]: Good callout 🙏

**Comment by brenbar** [comment]
@komamitsu Thanks for the review. All comments addressed.

**Comment by komamitsu** [comment]
@brenbar Thanks! LGTM :+1: 

After reviewing all the changes, `writeIntegerKeysAsStringKeys` feels a bit verbose? `supportIntegerKeys` (`false` by default) or something might be simpler. What do you think?

**Comment by brenbar** [comment]
@komamitsu I think your feedback is fair 😅 

Ready again for review 👍

**Review by komamitsu — APPROVED** [review]
LGTM, thank you!
